augur/io/sqlite3.py

```diff
@@ -0,0 +1,123 @@
+        Otherwise, a new connection is returned every time.
+        """
+        if self.use_in_memory_db:
+            return self.in_memory_db_connection
+        return sqlite3.connect(self.file, **connect_kwargs)
+
+    def execute(self, *args):
+        """Execute under a connection context."""
+        with self.connect() as con:
+            con.execute(*args)
+
+    def executemany(self, *args):
+        """Executemany under a connection context."""
+        # TODO: benchmark performance with cursor.executemany / con.commit at the end
+        with self.connect() as con:
+            con.executemany(*args)
+
+    def create_table(self, table_name:str, columns:List[str]):
+        """Create a table with all columns as TEXT type."""
+        self.execute(f"""
+            CREATE TABLE {sanitize_identifier(table_name)} (
+                {','.join(f'{sanitize_identifier(column)} TEXT' for column in columns)}
+            )
+        """)
+
+    def create_unique_index(self, table_name:str, column:str):
+        """Create a unique index on a column in a table."""
+        index_name = f'index_{table_name}_{column}'
+        self.execute(f"""
+            CREATE UNIQUE INDEX {sanitize_identifier(index_name)}
+            ON {sanitize_identifier(table_name)} ({sanitize_identifier(column)})
+        """)
+
+    def query_to_csv(self, query:str, path:str, header=True, chunksize:int=10000, **to_csv_kwargs):
+        """Query the database and write results to a tabular file.
+
+        Parameters
+        ----------
+        query
+            SQLite query string.
+        path
+            Path to the output file.
+        header
+            Write out the column names.
+        chunksize
+            Maximum number of rows to load into memory at once.
+        to_csv_kwargs
+            Additional keyword arguments passed to pandas.to_csv().
+        """
+        with self.connect() as con:
+            # Get results in chunks so they are not loaded into memory all at once.
+            df_chunks = pd.read_sql_query(query, con, chunksize=chunksize)
+
+            # Overwrite the file if it exists.
+            mode = 'w'
+
+            for df_chunk in df_chunks:
+                df_chunk.to_csv(path, mode=mode, header=header, **to_csv_kwargs)
+
+                # For subsequent chunks, use append mode and don't include the header.
+                mode = 'a'
+                header = False
+
+    def query_to_fasta(self, query:str, path:str, id_name:str, sequence_name:str, chunksize:int=10000):
+        with self.connect() as con, open(path, "w") as out_file:
+            # Get results in chunks so they are not loaded into memory all at once.
+            df_chunks = pd.read_sql_query(query, con, chunksize=chunksize)
+            for df_chunk in df_chunks:
+                sequences = df_chunk.apply(lambda row: SeqRecord(id=row[id_name], seq=Seq(row[sequence_name]), description=""), axis=1)
```

Comment:
Why use Pandas here?

Reply:
Good question. I took a stab at doing it without Pandas, and this is what it would be: 39997df7707d383a90976e563b26c5e5cd9744ff

In terms of development, there's pros and cons to both approaches.

- Pandas pros: I suspect this type of query→csv operation will be used frequently with various usages, and any extra features we may want down the road should be customizable though `to_csv_kwargs`. This would be easier to build on in comparison to managing connections/cursors to get the equivalent. Also, developers should be familiar with Pandas due to usage across the codebase.
- Pandas cons: need to read in chunks to avoid loading a large amount of data into memory.

Without Pandas, the pros/cons are inverted.

I did not try for sequences yet (`query_to_fasta()`). That would probably require a `csv.DictWriter` to access id and seq columns.

Note that I haven't tested performance differences, which should be an important consideration here.

Reply:
I think in Pandas there's a lot of underlying, hidden complexity waiting to bite us, and that it's really unnecessary for dumping a database query cursor to a file. Some of this is from experience being bitten.

I think overall the non-Pandas approach you drafted in https://github.com/nextstrain/augur/commit/39997df7707d383a90976e563b26c5e5cd9744ff is simpler and more understandable.

(Passing thru keyword args from a first-party function to a third-party function, as in `to_csv_kwargs`, is an anti-pattern, in my book, so it's not something I'd want to build on.)